DietPi-Software | Matrix Synapse Server

nonnorm opened the pull request:
<!--
Before submitting a pull request:
- Please ensure the target branch is "dev" (active development): https://github.com/MichaIng/DietPi/tree/dev
- Please ensure changes have been tested and verified functional.
-->
**Status**: Ready
- [x] Implement using PostgreSQL
- [x] Switch to building from source

**Reference**: https://github.com/MichaIng/DietPi/issues/4525

**Commit list/description**:
<!--
- DietPi-Config | Add "Fan control" option to "Performance Options"
![Screenshot](https://xxx.github.com/images/xxx.png)
- DietPi-Config | Add "Fan control" support for Odroid C2
- DietPi-Config | Syntax fix
-->
- Multiple | Add Synapse Server

nonnorm committed:
v7.4
+ Multiple | Add Synapse Server

nonnorm committed:
v7.4
+ DietPi-Software | Implement PostgreSQL usage

nonnorm committed:
v7.4
+ DietPi-Software | Add and remove G_EXEC

nonnorm committed:
v7.4
+ DietPi-Software | Allow you to set server name and analytics

nonnorm committed:
v7.4
+ DietPi-Software | Add G_EXEC_OUTPUT=1 to some commands

nonnorm committed:
v7.4
+ DietPi-Software | Restart service before trying to register new user

nonnorm committed:
v7.4
+ DietPi-Software | Change bind addresses to bind to the public network

nonnorm committed:
v7.4
+ DietPi-Software | Use buster-backports instead of official repo

nonnorm committed:
v7.4
+ DietPi-Software | Change way synapse is installed

nonnorm committed:
v7.4
+ DietPi-Software | Add missing path to G_CONFIG_INJECT

nonnorm committed:
v7.4
+ DietPi-Software | Add pre-made config file

Joulinar commented on dietpi/dietpi-software:
```diff
@@ -3617,6 +3642,16 @@ Package: openssl libssl*\nPin: origin packages.sury.org\nPin-Priority: -1' > /et
+			Banner_Installing
+
+			# APT package (have to install from backports until official repo supports ARM)
+			G_AGI -t buster-backports matrix-synapse
```
best to my knowledge, `buster-backports` source list is not available on Raspberry OS 32bit by default.

nonnorm replied:
Well, I'll have to build it from source then.

nonnorm committed:
v7.4
+ DietPi-Software | Build synapse from source

nonnorm committed:
v7.4
+ DietPi-Software | Fix paths

nonnorm committed:
v7.4
+ DietPi-Software | Fix uninstall

nonnorm committed:
v7.4
+ DietPi-Software | Fix bind_addresses

nonnorm committed:
v7.4
+ DietPi-Software | Replace G_CONFIG_INJECT with sed

nonnorm committed:
v7.4
+ DietPi-Software | Fix indentation

nonnorm committed:
v7.4
+ DietPi-Software | cd to synapse directory before creating config files

nonnorm committed:
v7.4
+ DietPi-Services | Rename synapse service

nonnorm committed:
v7.4
+ DietPi-Software | Remove unneeded config_path

nonnorm committed:
v7.4
+ DietPi-Software | Fix dietpi_userdata path

nonnorm committed:
v7.4
+ DietPi-Software | Re-add apparently needed --config-file

nonnorm committed:
v7.4
+ DietPi-Software | Fix typo

nonnorm committed:
v7.4
+ DietPi-Software | CodeFactor Fixes

nonnorm committed:
v7.4
+ DietPi-Software | Fix regex

nonnorm committed:
v7.4
+ DietPi-Software | Add piwheels dependencies

nonnorm committed:
v7.4
+ DietPi-Software | CodeFactor Fixes

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -2989,6 +3005,15 @@ _EOF_
+
+			Banner_Installing
+
+			G_AGI postgresql
```
Do we want to put the data directory into DietPi userdata, like we do for MariaDB? I guess a symlink will work just the same way.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -3617,6 +3642,22 @@ Package: openssl libssl*\nPin: origin packages.sury.org\nPin-Priority: -1' > /et
+
+			Banner_Installing
+
+			# Install dependencies for running on ARMv6/7 (piwheels) or building
```
There is an `x86_64` wheel as well, at least I think it's one: https://pypi.org/project/matrix-synapse/#files
I'll test, just leave the note to not forget.

MichaIng replied:
What are libopenjp2-7 libtiff5 libxcb1 for? I have none of them installed on x86_64 and build + service startup + web access work fine. Only psycopg2 needs to be compiled there (libpq-dev + gcc required), but matrix-synapse is pre-compiled.

nonnorm replied:
Whenever I tried to run it on RPi 32-bit, it gave me import errors until I installed those packages.  Strange, since none of them are listed on their piwheels project page.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -10811,6 +10852,73 @@ location = /.well-known/caldav  { return 301 /baikal/html/dav.php; }' > /etc/ngi
+
+			G_EXEC mkdir -p /mnt/dietpi_userdata/synapse
+
+			G_EXEC sudo -u postgres psql -c "create role synapse with login password '$GLOBAL_PW';"
```
Let's use a custom password here, a random one like we do in case of Nextcloud/ownCloud. As this is only for automated access from Synapse, there is no need to use the global software password.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -10811,6 +10852,73 @@ location = /.well-known/caldav  { return 301 /baikal/html/dav.php; }' > /etc/ngi
+				analytics='no'
+			fi
+
+			G_EXEC cd /mnt/dietpi_userdata/synapse
```
Don't forget to `cd /tmp/$G_PROGRAM_NAME` after done. Btw, I think creating and editing the config should only be done if it does not already exist, so that we do not touch or overwrite existing configs?

nonnorm committed:
v7.4
+ DietPi-Software | Fixes

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -15704,6 +15712,8 @@ _EOF_
 
 			Banner_Uninstalling
+
+			G_EXEC ln -s /mnt/dietpi_userdata/postgresql /var/lib/postgresql/11/main
```
On Bullseye, it will be `/var/lib/postgresql/13/main`, on Stretch it's `/var/lib/postgresql/9/main`, let's:
```sh
G_EXEC ln -s /mnt/dietpi_userdata/postgresql /var/lib/postgresql
```
And to avoid failures on reinstalls or when the package was manually installed before, we need to move the directory, if it exists already and remove the link, if it exists already. I'll push a commit.

MichaIng committed:
v7.4

+ DietPi-Software | PostgreSQL: Fix if-then syntax
+ DietPi-Software | PostgreSQL: Create database symlink to userdata as intended before installing the package, including migration of existing directories, and remove data on uninstall

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -10811,6 +10878,76 @@ location = /.well-known/caldav  { return 301 /baikal/html/dav.php; }' > /etc/ngi
+			local synapse_pass=$(tr -dc '[:alnum:]' < /dev/random | head -c30)
+
+			G_EXEC sudo -u postgres psql -c "create role synapse with login password '$synapse_pass';"
+			G_EXEC sudo -u postgres createdb --encoding=UTF8 --locale=C --template=template0 --owner=synapse synapse
```
Will this fail when the database exists already? In case we should skip it then.

MichaIng replied:
It fails.
The following can be used to check whether a role exists already:
```sh
sudo -u postgres psql -c '\du' | grep -q '^[[:blank:]]*synapse[[:blank:]]'
```
And the following could be used to check whether the database exists already:
```sh
sudo -u postgres psql -l -t | grep -q '^[[:blank:]]*synapse[[:blank:]]'
```

If the role exists already, I think we should change it's password instead and, if the config does already exist, re-apply it. It's a little nasty sed required to apply it to the correct place. I can push that commit.

The servername and analytics prompts, as well as register_new_matrix_user should be as well only called when the config file does not yet exist. The latter else fails.

nonnorm committed:
v7.4
+ DietPi-Software | Only create synapse database when it doesn't exist already

MichaIng committed:
v7.4

+ DietPi-Software | Synapse: Adjust build dependencies for x86_64
+ DietPi-Software | Synapse: Ask for server name, analytics and do register_new_matrix_user only on new installs
+ DietPi-Software | Synapse: Stop and remove service first

MichaIng committed:
v7.4

+ DietPi-Software | Syntax

MichaIng committed:
v7.4

+ DietPi-Software | PostgreSQL: Uninstall very late so that dependants are able to create backups etc
+ DietPi-Software | Synapse: Fix uninstall
+ DietPi-Software | Synapse: Update ARMv8 dependencies: Like on x86_64 only the psycopg2 module needs to be compiled, so that only a small set of build tools and libraries are required
+ DietPi-Software | Synapse: Create database role and database itself independently, only when missing. If the role is already present, update its password and likewise if the config file is already present, update the database password within.
+ DietPi-Software | Synapse: When asking for analytics, show "Yes" and "No" on buttons instead of "Ok" and "Cancel"

MichaIng committed:
v7.4

+ DietPi-Software | PostgreSQL: postgresql-common prevents all other packages from being auto-removed, hence select all of them explicitly

MichaIng committed:
v7.4

+ DietPi-Software | Synapse: Fix build on ARMv8 where make is additionally required to compile PyNaCl
+ DietPi-Software | Synapse: Merge install and config steps, as we want to do for all software titles
+ DietPi-Software | Enhance logging software description

MichaIng committed:
v7.4

+ README | Add Synapse sources to 3rd party list

MichaIng committed:
v7.4

+ DietPi-Survey_report | Add support for PostgreSQL with v7.4

MichaIng committed:
v7.4

+ CHANGELOG | Synapse: A Matrix homeserver implementation has been added with software ID 125.
+ CHANGELOG | PostgreSQL: A persistent advanced object-relational database server has been added with software ID 194.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -1073,6 +1073,15 @@ INDEX_BROWSER_TARGET=$INDEX_BROWSER_TARGET"
+		#------------------
+		software_id=125
+
+		aSOFTWARE_NAME[$software_id]='Synapse Server'
```
Shall we name it simply "Synapse"? There is no Synapse client, but general chat clients, like Element, so I think it's sufficient.

nonnorm committed:
v7.4
+ DietPi-Software | Rename Synapse Server to Synapse

MichaIng committed:
v7.4

+ DietPi-Software | PostgreSQL: Disable TCP/IP listener by default and instead assure that UNIX domain socket is enabled at expected location
+ DietPi-Software | Synapse: Start server in systemd unit via python3 call directly and omit synctl, which saves ~70 MiB memory and allows to pass custom command line parameters easily. Use this to load additional override configs from homeserver.yaml.d directory.
+ DietPi-Software | Synapse: Connect to PostgreSQL via UNIX domain socket, which provides ~30% faster queries.
+ DietPi-Software | Synapse: Apply database access details via override config, which is much safer than editing YAML with repeating and multi-line config blocks.
+ DietPi-Software | Synapse: Log to journal (journalctl -u synapse) by default and reduce log level to WARNING
+ DietPi-Software | Synapse: Switch to Type=notify, so that "systemctl start synapse" finishes only after the service is ready to listen. Else register_new_matrix_user can fail when the server was not yet ready. Taken from upstream: https://github.com/matrix-org/synapse/blob/develop/debian/matrix-synapse.service
+ DietPi-Software | Synapse: Implement quick reload with systemd unit, taken from upstream: https://github.com/matrix-org/synapse/blob/develop/debian/matrix-synapse.service

MichaIng committed:
v7.4

+ DietPi-Software | PostgreSQL: Remove possible left configuration files on uninstall

MichaIng committed:
v7.4

+ DietPi-Software | Synapse: Add forgotten service type change to Type=notify

MichaIng committed:
v7.4

+ DietPi-Services | Add PostgreSQL control

MichaIng committed:
v7.4

+ DietPi-Software | PostgreSQL: Fix override config creation

MichaIng committed:
v7.4

+ DietPi-Software | Synapse: Before checking and altering the database, assure that PostgreSQL is started, essential on reinstalls

MichaIng committed:
v7.4

+ DietPi-Software | Synapse: Clarify in comment a little why image codecs are required on ARMv6/7 (but not on ARMv8 and x64_64) 
+ DietPi-Software | Synapse: Error-handle most steps of the uninstall
+ DietPi-Software | Python 3: Remove obsolete pip cache on uninstall

MichaIng committed:
v7.4

+ DietPi-Software | Synapse: Since PyPI does not ship a Python 3.5 cffi wheel anymore, it needs to be compiled on Debian Stretch

MichaIng committed:
v7.4

+ DietPi-Software | PostgreSQL: Add a commit about why we don't disable the TCP/IP listener on Stretch
